src/category_theory/monoidal/Bimod.lean

```diff
@@ -0,0 +1,1020 @@
+
+/-- Left action for the tensor product of two bimodules. -/
+noncomputable
+def act_left : R.X ⊗ X P Q ⟶ X P Q :=
+begin
+  refine (preserves_coequalizer.iso (tensor_left R.X) _ _).inv ≫ _,
+  apply colim_map,
+  fapply parallel_pair_hom,
+  dsimp,
+  refine (𝟙 _ ⊗ (α_ _ _ _).hom) ≫ (α_ _ _ _).inv ≫ (P.act_left ⊗ 𝟙 S.X ⊗ 𝟙 Q.X) ≫ (α_ _ _ _).inv,
+  refine (α_ _ _ _).inv ≫ (P.act_left ⊗ 𝟙 Q.X),
+  { dsimp,
+    slice_lhs 1 2 { rw associator_inv_naturality },
+    slice_rhs 3 4 { rw associator_inv_naturality },
+    slice_rhs 4 5 { rw [←tensor_comp, middle_assoc, tensor_comp, comp_tensor_id] },
+    coherence, },
+  { dsimp,
+    slice_lhs 1 1 { rw id_tensor_comp },
+    slice_lhs 2 3 { rw associator_inv_naturality },
+    slice_lhs 3 4 { rw [tensor_id, id_tensor_comp_tensor_id] },
+    slice_rhs 4 6 { rw iso.inv_hom_id_assoc },
+    slice_rhs 3 4 { rw [tensor_id, tensor_id_comp_id_tensor] }, }
+end
+
+lemma id_tensor_π_act_left :
+  (𝟙 R.X ⊗ coequalizer.π _ _) ≫ act_left P Q =
+  (α_ _ _ _).inv ≫ (P.act_left ⊗ 𝟙 Q.X) ≫ coequalizer.π _ _ :=
+begin
+  erw map_π_preserves_coequalizer_inv_colim_map (tensor_left _),
+  simp only [category.assoc],
+end
+
+lemma one_act_left' : (R.one ⊗ 𝟙 _) ≫ act_left P Q = (λ_ _).hom :=
+begin
+  refine (cancel_epi ((tensor_left _).map (coequalizer.π _ _))).1 _,
+  dunfold X, dsimp,
+  slice_lhs 1 2 { rw [id_tensor_comp_tensor_id, ←tensor_id_comp_id_tensor] },
+  slice_lhs 2 3 { rw id_tensor_π_act_left },
+  slice_lhs 1 2 { rw [←monoidal_category.tensor_id, associator_inv_naturality] },
+  slice_lhs 2 3 { rw [←comp_tensor_id, one_act_left] },
+  slice_rhs 1 2 { rw left_unitor_naturality },
+  coherence,
+end
+
+lemma left_assoc' :
+  (R.mul ⊗ 𝟙 _) ≫ act_left P Q =
+  (α_ R.X R.X _).hom ≫ (𝟙 R.X ⊗ act_left P Q) ≫ act_left P Q :=
+begin
+  refine (cancel_epi ((tensor_left _).map (coequalizer.π _ _))).1 _,
+  dunfold X, dsimp,
+  slice_lhs 1 2 { rw [id_tensor_comp_tensor_id, ←tensor_id_comp_id_tensor] },
+  slice_lhs 2 3 { rw id_tensor_π_act_left },
+  slice_lhs 1 2 { rw [←monoidal_category.tensor_id, associator_inv_naturality] },
+  slice_lhs 2 3 { rw [←comp_tensor_id, left_assoc, comp_tensor_id, comp_tensor_id] },
+  slice_rhs 1 2 { rw [←monoidal_category.tensor_id, associator_naturality] },
+  slice_rhs 2 3 { rw [←id_tensor_comp, id_tensor_π_act_left, id_tensor_comp, id_tensor_comp] },
+  slice_rhs 4 5 { rw id_tensor_π_act_left },
+  slice_rhs 3 4 { rw associator_inv_naturality },
+  coherence,
+end
+
+end
+
+section
+
+variables [∀ X : C, preserves_colimits (tensor_right X)]
+
+/-- Right action for the tensor product of two bimodules. -/
+noncomputable
+def act_right : X P Q ⊗ T.X ⟶ X P Q :=
+begin
+  refine (preserves_coequalizer.iso (tensor_right T.X) _ _).inv ≫ _,
+  apply colim_map,
+  fapply parallel_pair_hom,
+  dsimp,
+  refine (α_ _ _ _).hom ≫ (α_ _ _ _).hom ≫ (𝟙 P.X ⊗ 𝟙 S.X ⊗ Q.act_right) ≫ (α_ _ _ _).inv,
+  refine (α_ _ _ _).hom ≫ (𝟙 P.X ⊗ Q.act_right),
```

Comment:
> Moreover, I prefer the definition written outside the begin-end block since it is more transparent. I'm not sure if it's a common understanding in mathlib though.

I'm not sure what exactly this means.  Staying in term mode when giving definitions and switching to tactic mode when providing proof terms?  Or perhaps extracting proof obligations into lemmas?

Reply:
I meant the former. I acknowledge the latest commit about this comment.